Try improving the side UI for nested blocks

jasmussen opened the pull request:
This PR leverages the recent block margin refactor to try and improve how nested blocks UI works.

This is a first step, with many possible steps to take in the future, but from a high level it seeks to make immediate improvements to:

- Reducing the visual weight of nested blocks UI, including the dimensions & paddings
- Making the editor appearance look as much the same as the frontend, as possible

![side ui improvements](https://user-images.githubusercontent.com/1204802/37518360-166234bc-2915-11e8-9fcd-6a044a541336.gif)

It's still very much a work in progress, but even now it has a lot of benefits over master:

- Because the block UI is stacked, the block dimensions are now no longer marred by the excessive margins we used to make the side UI work
- The side UI has a background (experimental) to make it visible even when it covers content

There are still many more todos (I won't list them, some of them will be obvious), but before I go too much farther I would love some thoughts on this implementation as it stands. 

If the approach has merit, then there is one additional change we should make in this PR before it's ship ready, and that is to ensure that the _parent block is easily selectable_. One way to do this could be to _always show the gray borders around the parent block, when interacting with a child block_. An alternative would be to always show a side-UI type-icon in the vicinity of the parent block, which when clicked selects the parent. This could also serve as a drag handle, now that the side UI is stacking. 

Additional ideas for future enhancements could be to _dim_ adjecent blocks when a child block is selected. It would also be nice to look at improving how hovering the parent block works. It can be hard to _unhover_ as it stands.

---

The key point of this PR is to address the side UI and how it affects nested blocks. The approach outlined in this PR was chosen because it was the simplest of them all, and requires the least UI rejiggering to work. 

If for whatever reason this approach isn't feasible after all, an alternative approach we've discussed involves surfacing the mobile UI for nested contexts, which would look something like this:

![plan b](https://user-images.githubusercontent.com/1204802/37518320-f3c2cf7a-2914-11e8-9517-94161cb13b88.png)

This UI will still definitely be surfaced on phone breakpoints, but the downside it has is that it increases the height of the selected blocks, which presumably can feel jumpy.

---

Here is some demo markup you can use to test the columns block:

```
<!-- wp:paragraph -->
<p>Loem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus.</p>
<!-- /wp:paragraph -->

<!-- wp:columns -->
<div class="wp-block-columns has-2-columns">
    <!-- wp:paragraph {"layout":"column-1"} -->
    <p class="layout-column-1">Lorem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus.</p>
    <!-- /wp:paragraph -->

    <!-- wp:paragraph {"layout":"column-1"} -->
    <p class="layout-column-1">Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no.</p>
    <!-- /wp:paragraph -->

    <!-- wp:paragraph {"layout":"column-1"} -->
    <p class="layout-column-1">Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</p>
    <!-- /wp:paragraph -->

    <!-- wp:paragraph {"layout":"column-2"} -->
    <p class="layout-column-2">Lorem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus.</p>
    <!-- /wp:paragraph -->

    <!-- wp:paragraph {"layout":"column-2"} -->
    <p class="layout-column-2">Eum cu unum choro albucius.</p>
    <!-- /wp:paragraph -->

    <!-- wp:paragraph {"layout":"column-2"} -->
    <p class="layout-column-2">Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri.</p>
    <!-- /wp:paragraph -->
</div>
<!-- /wp:columns -->

<!-- wp:paragraph -->
<p>Sea id autem nominavi deseruisse. Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has.</p>
<!-- /wp:paragraph -->
```

Commit:
Improve nested blocks UI

This PR is very much a work in progress. Pushing now early, so I can work on it both at work and at home. It will not even be made a PR yet, and with a good squash there's every chance you won't even read this message. Github squashing is like Terminator time travel!

jasmussen commented:
As a sidenote, enhancements made to the side UI in this PR overlaps and will also benefit #5400 and #4223.

jasmussen commented:
Okay, I've pushed a lot of fixes to this, and and the nested UI (tested on columns block) is very much improved in this branch compared to master. Here's a recent video:

![latest video](https://user-images.githubusercontent.com/1204802/37650822-6b3fe744-2c36-11e8-9007-60fd86494d33.gif)

The UI is basically solid, except for one thing. In https://github.com/WordPress/gutenberg/pull/5658/commits/a943c230e38d82c73bb44bedda0c6ba4e0207b39, I disabled the ability to hover or even select the parent block (in this case the columns block). Because there is only a tiny sliver of pixels above and below the columns block, the margin essentially, where can hover or select the block. This area will be even smaller, virtually non existant, when we enable margin collapsing. It wasn't working very well at all, so I disabled it pending an explicit action to select the parent block. 

Here are mockups for how to enable that to happen. When you've a nested block selected, you see this:

![v3 nested block selected](https://user-images.githubusercontent.com/1204802/37651171-8a61f6c0-2c37-11e8-8ed8-9d4494783ef1.png)

You can then click the "Quote" icon in the editor bar to move up the breadcrumb trail and select the parent block:

![v3 parent block selected](https://user-images.githubusercontent.com/1204802/37651213-af349e62-2c37-11e8-8754-f260f1f2423b.png)

Note that in the mockups above, there is no switcher in child blocks, only the parent can be "switched". This can definitely be revisited in the future, but by limiting transformations to the parent block _for now_ we don't have to deal with what happens when you transform a paragraph inside a quote block into a quote (nested quote). 

Also, the side UI, while bigger and easier to hit in this branch, has additional improvements coming in #5400. 

---

Because this PR is in a fairly good place, and because the code is relatively simple, I would recommend the next step is that someone with a little time on their hands (@mtias? @aduth? @youknowriad ) can help implement a way to select the parent block, perhaps the breadcrumb trail mocked up above. Then we can merge this in, and look at making additional enhancements in future PRs.

My biggest worry is that this PR balloons and becomes too complex.

aduth commented:
I'll spend some time iterating at a parent selector.

jasmussen commented:
Note that I think @mtias had some ideas he wanted to try too, possibly related to this.

aduth committed:
Block List: Show insertion point prior to block

Better compatibility with high z-index of sibling inserter conflicting with hover behavior of block.

aduth committed:
Block List: Show block breadcrumb prior to block

aduth commented:
Rebased to resolve conflicts (squashed in the process). Pushed some (rather incomplete) iterations on a hover toolbar to select block or parent block. Will continue tomorrow, but sharing what I have in case you were itching to start hacking away on (much-needed 😄) design improvements.

jasmussen commented:
Thank you Andrew, really appreciate that. Pushed some fixes:

![nested](https://user-images.githubusercontent.com/1204802/37700066-6e9c98d8-2cea-11e8-9929-c1e8850ce800.gif)

This is based on a different mockup for the hover element:

![screen shot 2018-03-21 at 09 28 21](https://user-images.githubusercontent.com/1204802/37700076-7c38e2da-2cea-11e8-9ba0-ba4698693f2a.png)

It may be worth pushing this soon, so we can refine separately. Thoughts?

aduth commented:
I may revise (simplify) of the implementation if the button is really only meant to navigate one level up to the immediate parent. As-is, if there was double nesting, it would display two buttons, both with "Select parent block" tooltips. Probably not expected, and instead the user, if wanting to navigate upward through to the grandparent block, would select the single "Select parent block" one at a time.

aduth commented:
Rebased again to resolve conflicts, and updated implementation. On that end, it feels complete.

I noticed an issue though, which is that with new styling, it's now impossible to add content to the second column in a newly-created columns block.

aduth commented:
In fact, if you add a Columns block and immediate deselect it, it's impossible to interact with it at all.

aduth committed:
Editor: Expose BlockTitle component

jasmussen commented:
Thank you for working on this. I'm having to wrangle a baby right now but can look later, I'm guessing that it's the pointer events handling acting up on empty columns.

aduth commented:
Yes, the `pointer-events: none` has many consequences. One also cannot use block movers or ellipsis menu on a selected columns block, or use the sibling inserter shown in the space between the Columns block and its prior block.

jasmussen commented:
> Yes, the pointer-events: none has many consequences. One also cannot use block movers or ellipsis menu on a selected columns block, or use the sibling inserter shown in the space between the Columns block and its prior block.

The thing I was trying to accomplish by adding that rule, was to prevent hovering of the _columns_block. You can try commenting it out in https://github.com/WordPress/gutenberg/pull/5658/files#diff-80732b9fe9aee13610c0e0b391943f2aR13

the reason being, the only hoverable area for the parent block is a few pixels above and below the entire thing. Which means if you move the mouse from up before the columns block and onto a child block, `is-hovered` is applied to the columns block and will only be removed if you carefully hover upwards again and hit those few pixels.

Perhaps a different approach, simply, is to _allow_ this to happen, and then just _remove_ `is-hovered` from the parent block, whenever a child block is hovered. 

This might also allow you to select the parent block by clicking an empty column?

jasmussen commented:
In this video, I gave the columns block a red background, and the child blocks green backgrounds. Note how hovering the red area invokes `is-hovered` on the columns block, and similarly when hovering the green area for child blocks. 

![hover](https://user-images.githubusercontent.com/1204802/37761107-065b4af6-2db9-11e8-90b3-fe97e8f6961b.gif)

The problem is, if you first hover the red area, and then move over the green area, `is-hovered` is not unset on the red block. Meaning if you move from red, to green, to white, the red highlit block remains hovered, even though the mouse is nowhere near it.

This is what I tried to fix using the pointer-events hack, but I'm going to undo that as it's clearly not ideal. On the other hand, it _is_ desirable to be able to _click_ the red area to select the parent block, when for example the right column is not filled out. 

@aduth do you think it would be feasible, and in scope of this PR, to _unset_ is-hovered on the parent block when hovering a child block?

Commit:
Undo the pointer-events trick.

aduth commented on blocks/library/paragraph/editor.scss:
```diff
@@ -2,6 +2,11 @@
 	background: white;
 }
 
+// Don't show white background when a nesting parent is selected
```
What's the reasoning for the white background in the first place?

jasmussen replied:
I actually can't recall. Hmm. @youknowriad?

jasmussen replied:
Looks like it was introduced here: https://github.com/WordPress/gutenberg/commit/775428e4778f406f1b8c815b54518d2350edc49c#diff-c2685bc3733aece65a4ed211668982e0R2

So it's probably related to the ability to colorize paragraph backgrounds. But removing it doesn't seem to make a difference. @jorgefilipecosta any thoughts?

aduth replied:
Seems we shouldn't want to assume a white background, especially if the plan is to [support themes to define their own editor styles](https://youtu.be/XOY3ZUO6P0k?t=3032).

jorgefilipecosta replied:
The reason why we set a white background is related to the contrast verification. If the background was never set getComputedStyles returns back background (if I remember correctly). So adding a background fixes the problem. If themes change the background it also works well because as long as a background is set we are able to retrieve it. The problem happens when no background is set. As we get back background color when background is not set we cannot differentiate and assume is white background because in fact it may be set to black. I will try to research this in more detail maybe there is a way to detect when the background is not set and in that case assume white.
Replicate from comment https://github.com/WordPress/gutenberg/pull/5273#discussion_r176880358.

aduth replied:
But this only works if the style is explicitly set, i.e. a theme would need to target and override the `.editor-block-list__block:not( .is-multi-selected ) .wp-block-paragraph` selector (or selector of equivalent specificity). And the contrast checker would otherwise falsely report contrast issues on the basis of a white background? i.e. if theme actually had black background where a white text would be very contrasting, it would be reported by the editor as an issue.

I'm wondering if it's really worth the trouble to assume a default value (vs. disabling until a background color is known / set).

Alternatively, could we recurse up through parent elements until we find one where background color is not the default value?

jorgefilipecosta replied:
I think we can remove the background color, for now, as it is causing us troubles so we can advance.
After, I really like the idea, of recursively iterating up to the parent elements. And I will try to explore it. If we find it not a good solution than we disable contrast checker for colors not explicitly set.

aduth commented:
>@aduth do you think it would be feasible, and in scope of this PR, to unset is-hovered on the parent block when hovering a child block?

I'll take a look at this. It may be tricky because of the way events are handled / ignored between parent blocks and their children.

aduth commented:
This may have been easier with the original hover state which was changed in #5078, since previously it was the case that at most one block could be hovered at a time.

aduth committed:
Block List: Cancel hover on handled descendant mouseover

aduth commented:
>>@aduth do you think it would be feasible, and in scope of this PR, to unset is-hovered on the parent block when hovering a child block?
>
>I'll take a look at this. It may be tricky because of the way events are handled / ignored between parent blocks and their children.

This should be working now in 0d0c8e0.

There's an end-to-end test failure that looks legitimate. Investigating now.

aduth commented:
Rebased once more. The E2E test failure was due to the fact that we assumed the sibling inserter was positioned after a block, but this was moved in 6203d2c and the test needed to be updated to accommodate (the rebased commit does so).

aduth commented:
Hmm, noticing some weirdnesses with the hover behavior for shared nested blocks (notably, it is difficult to access the "More Options" menu for the shared block of a columns block).

aduth commented:
Last observed issue is introduced here, but not entirely relevant. Created a separate fix at #5748.

aduth commented:
Rebased to resolve conflicts. Aside from a failing E2E test from the new Puppeteer setup, which I'll address in the morning, I'm not aware of other work that needs to be done here.

mtias commented:
@aduth if you switch to render the block toolbar at the top via ellipsis menu setting, we should probably render the label at the top as well. @jasmussen would you agree?

mtias commented:
@jasmussen also, there's a small height jump when you hover the separator and pullquote blocks (can check in demo post).

jasmussen commented:
> if you switch to render the block toolbar at the top via ellipsis menu setting, we should probably render the label at the top as well. @jasmussen would you agree?

Yes, I agree, good catch!

paulwilde commented:
I just tested this out and found a minor visual bug, but otherwise a huge improvement over the prior UI. I believe some of these are Safari specific.

Large border on the down arrow. Fixed by adding `border-width: 1px` onto `.editor-block-list__layout .editor-block-list__layout .editor-block-mover__control`. 

<img width="468" alt="screen shot 2018-03-27 at 19 12 10" src="https://user-images.githubusercontent.com/704594/37986218-cfaa70f2-31f2-11e8-8516-8725a5fb3869.png">

Also worth mentioning, however I doubt this is the ticket where it'll be solved - In the columns block, if you are focussed one column to add content and then hover over the adjacent column, then the 'Add block' button is shown on top of the 3 'quick add' items.

<img width="845" alt="screen shot 2018-03-27 at 19 19 14" src="https://user-images.githubusercontent.com/704594/37986588-c4255746-31f3-11e8-9108-e14ed6a28758.png">

Maybe a better UI for the initial state of the columns would be to disable the 3 'quick add' menu, and then have a centered plus icon similar to the one on the gallery block:

<img width="804" alt="screen shot 2018-03-27 at 19 27 02" src="https://user-images.githubusercontent.com/704594/37986970-e4c66598-31f4-11e8-82b1-3926c80fec0c.png">

jasmussen commented:
Good call @paulwilde, thanks for testing this. 

I will revisit the side UI in general, as part of https://github.com/WordPress/gutenberg/issues/5400, and may try to reintroduce the circleshaped buttons depending on how that works. But until that time it seems like a nice fix to add that border trick. 

Also agree on the appender — this also needs love, and would be good to revisit separately. Essentially we need the appender to behave as if it was in a responsive breakpoint, just inside columns as well. 

By the way are those screenshots showing an editor style? ✊ 🏅  — completely unrelated to this PR, but I'd love for you to write down your biggest challenges in making editor styles work. This PR is definitely designed to make it easier for columns, and I hope to improve things further by allowing margins to collapse shortly after. But curious what other things you'd like to see improved to make editor styles easier.

paulwilde commented:
@jasmussen Yeah I've just started building a production site in Gutenberg (Simple 3-4 page website, so not too concerned if there's API changes when Gutenberg is merged). The site itself is quite complex though, so I couldn't imagine not building it in something like Gutenberg without comprising the level of control for the client.

I've added a single stylesheet which gets enqueued into `enqueue_block_editor_assets`, which is basically a separate stylesheet generating the equivalent styles for Gutenberg based on the front-end stylesheet. When I make updates to my front-end styles, those are then reflected in Gutenberg without the hassle of maintaining 2 separate stylesheets too much. I've also tweaked the width of the block container so that it matches the frontend 1 for 1. Also when I update the width of the front-end container its then reflected in Gutenberg (Yay, variables).

The only thing that would have been nice, but I ran into a dead-end was adjusting the background colour of the canvas, as with that it would be truly 1 for 1 to the front-end. Basically there's some hard-coded white backgrounds in the UI, and also the surrounding side UI (Movers, Block containment border, Add block + icon, 3 quick add icons) are all in hex greys so they didn't look too nice against the background of this site (`#fff4dc`). If they were using the `rgba()` equivalents on the white background, then they would actually look fine on different coloured backgrounds I think.

I've ran into a couple of other limitations from actually creating a production site as well, but one of note is `add_theme_support( 'editor-color-palette' )` being too limited, specifically with it just being globally added to any block and not having fine-grained control of which colours target which block. I'm currently looking at writing a custom implementation for this site which sets the colour schemes on a per-block basis (And using classes, instead of inline styles). I'm also not a fan of the text colour being set independently of background colours, I prefer the text colour being set based on the background.

I'm looking at implementing something like #4900 also, however I believe this might require the `InnerBlocks` component from #5452 to implement. One requirement for this site is also to have sections where the background of each section is flush with the sides of the screen, so basically the background is outside of the main 'block canvas'. I'm not sure if that'll be possible outside of Gutenberg, I'll find out soon enough.

Commit:
Fix arrow border and pixel shift.

Fix issue with nesting when only one block present on the movers. Props @paulwilde.

Fix issue with separator and pullquote pixelshifting on hover. Turns out this was due to collapsing margins, and iffy rounding that happened when a new toolbar was present. For now I prevented margins from collapsing on those, but this should be revisited and refactored separately.

jasmussen commented:
Pushed a fix for the small jump mentioned in https://github.com/WordPress/gutenberg/pull/5658#issuecomment-376374872 and the pixel in https://github.com/WordPress/gutenberg/pull/5658#issuecomment-376627614.

What remains for us to ship this? It would be nice to have in the next release if we can. So far as I can see:

- Fix the failing tests

Moving the hover label to be in the top toolbar when "Fix toolbar to top" is checked, should that be part of _this_ PR, or can that wait?

Future improvements include:

- Improving the appender in nested blocks, so it shows its responsive UI there
- Allow margins to collapse refactor
- Improve side UI overall as part of #5400

aduth commented:
>@aduth if you switch to render the block toolbar at the top via ellipsis menu setting, we should probably render the label at the top as well.

Hmm, I'm not sure. I thought one of the goals of the label was to serve as a reliable target for selecting the block. If the label is rendered in the top toolbar, it is impossible to both hover the block and click the label to select it.

jasmussen commented:
> Hmm, I'm not sure. I thought one of the goals of the label was to serve as a reliable target for selecting the block. If the label is rendered in the top toolbar, it is impossible to both hover the block and click the label to select it.

That's a solid point, and certainly solid enough, I think, to suggest we should move on with this PR without that change, and then revisit based on feedback.

aduth commented:
This failing test is actually very obscure. It's a legitimate error. You can reproduce if you use the between-inserter at a point where the toolbar would appear if the block were selected (slightly _above_ the block, more on the left edge). I'm led to believe this is actually a very particular edge case where, following [the specification](https://www.w3.org/TR/uievents/#event-type-click) which details that the click event only occurs on the top-most element at the time of pointer release, the insertion point button does not fire the `onClick` because in the meantime, the block has received a `focus` event which has caused the contextual toolbar to display. Thus, at the time of pointer release, the topmost element is the contextual toolbar. This was not an issue previously only because the insertion point occurred at the end of a block, so while the block was still selected, since its selected toolbar appeared above the block, it did not overlap the insertion point at the time of pointer release.

aduth commented:
Thinking some more, I don't think the sibling inserter should be rendered within a block. I know that at one point it was not, and it was later changed, but conceptually it ought not really be a descendent. Particularly with this focus behavior, we want it such that any focus within a block should trigger it to become selected. This is true for the block mover buttons and the more options menu (contained in the chrome surrounding the block's editable form).

But we _don't_ want this behavior for the sibling inserter. You can see even in the master / latest release that the block briefly becomes selected when clicking the sibling inserter. This is because it receives focus but then the inserted block becomes selected.

To me, it seems the ideal DOM / virtual hierarchy position for these inserters is _between_ the blocks, not within them.

jasmussen commented:
> To me, it seems the ideal DOM / virtual hierarchy position for these inserters is between the blocks, not within them.

Interesting. Should this be fixed as part of this PR, or addressed separately?

jvisick commented:
@paulwilde That's interesting to here about your thoughts on building out a site with custom styles. We are going through a similar process on a current project in seeing how far we can get with Gutenberg. Currently we have a custom section block using the `InnerBlocks` component that is functional (and that goes fullwidth to the edge of the screen). Eventually I want to push it up as a PR for #4900. If you have thoughts for how a section type block should work I encourage you to add them to the discussion on that issue!

jvisick commented:
@jasmussen Agreed, it's on the list...

jvisick commented:
@jasmussen I ran into this while testing with the column block and setting the  alignment to full width. The up/down blocks are overlapping the block controls, and the content is cut off by the edge of the editor box.

![screen shot 2018-03-28 at 2 43 03 pm](https://user-images.githubusercontent.com/2252216/38058151-b3419d24-3296-11e8-9542-eab142cba436.png)
![screen shot 2018-03-28 at 2 43 22 pm](https://user-images.githubusercontent.com/2252216/38058154-b5b14848-3296-11e8-9e4f-50b9c3a93ee3.png)

aduth commented:
>Interesting. Should this be fixed as part of this PR, or addressed separately?

Well, if we're content to live with the bug that sometimes prevents sibling block creation for short-term, we could address separately.

It's not possible to revert back to having the sibling inserter rendered after the editable display though, since it conflicts with hover behavior (this can be seen locally by applying `git revert bd92e42cb612424d58aef21f5a4044068b67ddad` to your copy of the branch).

I did some brief exploring of hacky resolutions to no immediate avail (like the Button `onFocus={ stopPropagation }`).

Commit:
Fix issues with wide and fullwide

This fixes a few issues with wide and fullwide alignments. Props for reporting, @jvisick.

- Movers now work on wide and fullwide, and look good
- Hovering now both works without horizontal scrollbars on wide items, and they aren't overlapped by the movers
- Wide alignment blocks now don't have negative margin to accommodate side UI.

jasmussen commented:
Thank you for the report @jvisick, I pushed a few fixes.

> Well, if we're content to live with the bug that sometimes prevents sibling block creation for short-term, we could address separately.

I'd personally be, but I also don't want us to create technical debt. If time isn't an issue, what would the best path forward be?

jvisick commented:
@jasmussen This is improved 😃 , but it looks like the movers on the first column and the settings button on the last column are no longer visible.

![col-ui-2](https://user-images.githubusercontent.com/2252216/38105652-79b96be4-3341-11e8-980c-862063e7d669.png)

![col-ui-1](https://user-images.githubusercontent.com/2252216/38105651-799ae070-3341-11e8-9e28-e89aae374520.png)

I tried adjusting the position of the ui to be just inside the blocks but then they cover up the text/content you are trying to edit. Maybe they need to go to the top, like the mover does for the parent column block?

![screen shot 2018-03-29 at 11 06 51 am](https://user-images.githubusercontent.com/2252216/38105663-80b737a0-3341-11e8-8baf-c85010c751b6.png)

Also, and this may be better addressed elsewhere, but I find that having the block border show up on hover is an improvement in navigating/selecting blocks. It's especially helpful with multiple nested blocks. I have been adding that as a custom style for our theming.

aduth commented:
After staring at this focus issue for far too long, and realizing that moving the sibling inserter's markup to between blocks would cause its own separate set of issues (loss of hover on the block), and with consideration of potential iterations on the sibling inserter via #5396, I'm inclined to simply disable the failing E2E test for now and move forward with design changes, addressing between inserter issues separately ahead of the next release.

In addition to above notes, I have found a few of my own quibbles:

_Columns child default inserter overlaps with columns movers:_

![image](https://user-images.githubusercontent.com/1779930/38151159-1e65d12a-3430-11e8-9b24-b22089373436.png)

_Likewise, second column default appender overlaps recent items from first column:_

![image](https://user-images.githubusercontent.com/1779930/38151222-6696a140-3430-11e8-9c17-08a972105712.png)

Commit:
Improve side UI for fullwide.

Props for bugtesting, @jvisick

This adds padding left and right on fullwide columns. We can revisit this in the future.

Commit:
Improve the appender inside columns, so it's the same UI as it is on mobile.

jasmussen commented:
In https://github.com/WordPress/gutenberg/pull/5658/commits/d60f648622527b252404aba390259cb6cdd41fc3, I fixed issues noted by @jvisick, thank you! This hardened the CSS a bit, and it's now nicer. Movers work even for child elements in columns now. Though I left a few notes in a ToDo for how we can enhance this in the future.

---

I also pushed some improvements to issues noted in https://github.com/WordPress/gutenberg/pull/5658#issuecomment-377605262, in https://github.com/WordPress/gutenberg/pull/5658/commits/7c8b7becbe36cae2254b5a23ec2a2568e52fbf27. Pushed this as a separate commit, as the changes there are slightly more questionable. I mean, it looks good, or at least better:

![appender ui](https://user-images.githubusercontent.com/1204802/38186554-eb3cbee6-3653-11e8-8e25-1ceda7381779.gif)

☝️ that's the mobile UI inside columns. We should also remove the "Write your story" so it isn't there in columns, but at least now it's the right font. 

The reason it's slightly questionable is that these styles are written in the editor CSS for the columns block, and aren't part of the appender component itself, which would be the ideal, and better scalable solution. I suspect we'd need to have access to [element queries](https://github.com/tomhodgins/element-queries-spec), in order to make that happen, though — in fact lots of UI could be improved if we had access to that. 

---

Depending on your thoughts, @aduth, I think we should merge this in as soon as we can, so it doesn't stay in a branch for much longer. It brings some much needed improvements to the block UI itself, and it'd be good to get this out there.

aduth committed:
Testing: Click at rightish of between inserter

aduth commented:
Rebased to resolve conflicts.

Pushed f67884c to trick the E2E tests into passing by not clicking at the problematic area of the inserter, noting that this is a temporary resolution.

Will do a quick review of the code, but otherwise agree we should get this in soon and iterate.

aduth commented on editor/components/block-list/style.scss:
```diff
@@ -334,23 +326,37 @@
+	&.is-hovered {
+		.editor-block-settings-menu,
+		.editor-block-mover {
+			z-index: 9999;	// @todo
```
Was this planned to be addressed here?

jasmussen replied:
Ack yes that was intended. All it needs is a proper z index with the mixin. I'm out right now but can look later, alternately feel free to push a fix if you have bandwidth.

aduth replied:
There's a `'.editor-block-mover': 1` z-index as well. Is that one made redundant with this? Can the controls always have a high z-index, even without the `is-selected` / `is-hovered` qualifiers? Are they even present when one of those conditions doesn't apply?

In meantime, I've pushed e1336d3 with reference to `z-index` list here.

jasmussen replied:
I believe that as I worked on this, I decided that yes, the side UI can always be above other things, except in the mobile configuration. So I'd think the old one is redundant now.

aduth committed:
Block List: Assign mover z-index via map entry

jasmussen commented:
@aduth was there anything left to fix before merging?

jasmussen commented:
🔥🎉❤️❤️

aduth commented:
We should remove the redundant `z-index`.

aduth commented:
Noticing an issue: When multi-selecting within columns, it's impossible to click on the "More Options" button:

![image](https://user-images.githubusercontent.com/1779930/38308069-e90eb34a-37e3-11e8-93c7-e262b94a9728.png)

(Keyboard behaviors appear unaffected)

jasmussen commented:
Going to take a look, thanks.
